# Atheris Command Line Options

## Run with coverage.py

see [05_atheris_code_coverage.ipynb](./05_atheris_code_coverage.ipynb)



## Basic Usage

In [17]:
!pwd
!python demo.py -atheris_runs=10000000 -max_len=100

/home/user/lectures/atheris/learn


INFO: Using built-in libfuzzer
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 1475448587
INFO: A corpus is not provided, starting from an empty corpus
#2	INITED cov: 2 ft: 2 corp: 1/1b exec/s: 0 rss: 65Mb
#20	NEW    cov: 4 ft: 4 corp: 2/5b lim: 4 exec/s: 0 rss: 65Mb L: 4/4 MS: 3 InsertByte-CrossOver-InsertByte-
#2097152	pulse  cov: 4 ft: 4 corp: 2/5b lim: 100 exec/s: 1048576 rss: 65Mb
#4194304	pulse  cov: 4 ft: 4 corp: 2/5b lim: 100 exec/s: 1048576 rss: 65Mb
#8388608	pulse  cov: 4 ft: 4 corp: 2/5b lim: 100 exec/s: 1048576 rss: 65Mb
Done 10000000 in 10 second(s)


#### 参数：

- -atheris_runs: 运行的次数
- -max_len: 输入字节的最大长度
- -max_total_time: 最大运行时间（秒）


#### 运行结果解析：

**INFO: Using built-in libfuzzer**
- **含义**：表示正在使用内置的 libFuzzer。


**WARNING: Failed to find function "__sanitizer_acquire_crash_state".**
- **含义**：这些警告表示未找到某些函数，但这通常不会影响模糊测试的运行。这些函数通常与地址消毒器（AddressSanitizer）相关，用于在崩溃时获取更多调试信息。


**INFO: Running with entropic power schedule (0xFF, 100).**
- **含义**：表示正在使用熵调度策略（entropic power schedule），这是 libFuzzer 的一种输入选择策略，用于更有效地探索输入空间。


**INFO: Seed: 1475448587**
- **含义**：表示使用的随机种子。这个种子用于初始化随机数生成器，以确保模糊测试的可重复性。


**INFO: A corpus is not provided, starting from an empty corpus**
- **含义**：表示没有提供初始语料库，模糊测试从空语料库开始。这意味着模糊测试将从头开始生成输入数据。


**#2 INITED cov: 2 ft: 2 corp: 1/1b exec/s: 0 rss: 65Mb**
- **#2**：表示这是第 2 次执行。
- **INITED**：表示初始化完成。
- **cov: 2**：表示覆盖率为 2。
- **ft: 2**：表示特征数为 2。
- **corp: 1/1b**：表示语料库包含 1 个输入，总大小为 1 字节。
- **exec/s: 0**：表示每秒执行次数为 0（初始化阶段）。
- **rss: 65Mb**：表示驻留集大小（内存使用量）为 65MB。

**#20 NEW cov: 4 ft: 4 corp: 2/5b lim: 4 exec/s: 0 rss: 65Mb L: 4/4 MS: 3 InsertByte-CrossOver-InsertByte-**
- **#20**：表示这是第 20 次执行。
- **NEW**：表示发现了新的覆盖率。
- **cov: 4**：表示覆盖率增加到 4。
- **ft: 4**：表示特征数增加到 4。
- **corp: 2/5b**：表示语料库包含 2 个输入，总大小为 5 字节。
- **lim: 4**：表示当前输入长度限制为 4 字节。
- **exec/s: 0**：表示每秒执行次数为 0。
- **rss: 65Mb**：表示驻留集大小为 65MB。
- **L: 4/4**：表示输入长度为 4 字节。
- **MS: 3 InsertByte-CrossOver-InsertByte-**：表示使用了 3 种变异策略：插入字节、交叉变异、插入字节。

**#2097152 pulse cov: 4 ft: 4 corp: 2/5b lim: 100 exec/s: 1048576 rss: 65Mb**
- **#2097152**：表示这是第 2,097,152 次执行。
- **pulse**：表示这是一个周期性输出，用于显示当前状态。
- **cov: 4**：表示覆盖率为 4。
- **ft: 4**：表示特征数为 4。
- **corp: 2/5b**：表示语料库包含 2 个输入，总大小为 5 字节。
- **lim: 100**：表示当前输入长度限制为 100 字节。
- **exec/s: 1048576**：表示每秒执行次数为 1,048,576。
- **rss: 65Mb**：表示驻留集大小为 65MB。

**Done 10000000 in 10 second(s)**
- **含义**：表示模糊测试在 10 秒内完成了 10,000,000 次执行。